File: app/src/main/java/com/kmt/healthanalyzer/domain/report/ReportModel.kt
package com.kmt.healthanalyzer.domain.report

import kotlinx.serialization.Serializable

/**
 * Contrat de données du rapport de santé.
 *
 * Ce modèle est la **seule** frontière entre le calcul et le dessin. Le moteur de
 * graphiques est écrit une fois en SVG (`web/report/`) ; il est utilisé tel quel par la
 * version web et par la WebView Android. Les deux plateformes doivent donc produire
 * exactement le même JSON, sinon les deux versions divergent.
 *
 * Règles à respecter en modifiant ce fichier :
 * - Toute modification ici doit être reportée dans `web/lib/report-model.js`.
 * - Les dates sont des chaînes ISO `AAAA-MM-JJ`, les mois des chaînes `AAAA-MM`.
 * - Les heures relatives sont exprimées en heures décimales autour de minuit :
 *   `-1.5` vaut 22h30, `2.25` vaut 2h15. Ce choix permet de tracer une nuit sur un axe
 *   continu, sans coupure à minuit.
 * - Une valeur absente vaut `null`, jamais zéro. Un zéro se dessine ; un trou ne se
 *   dessine pas.
 *
 * Ce modèle ne quitte jamais l'appareil. Il porte des mesures quotidiennes, donc il ne
 * doit jamais être envoyé à un fournisseur de LLM. La seule porte de sortie reste
 * `HealthPromptBuilder`, qui n'envoie que des agrégats hebdomadaires.
 */
@Serializable
data class ReportModel(
    val meta: ReportMeta,
    val tiles: List<ReportTile>,
    val sleep: SleepSection,
    val heart: HeartSection,
    val activity: ActivitySection,
    val body: BodySection,
    val stress: StressSection,
    val breathing: BreathingSection,
    val correlations: List<CorrelationItem>,
    val narrative: ReportNarrative? = null,
)

/** En-tête du rapport : période couverte et volumétrie, pour situer la fiabilité des chiffres. */
@Serializable
data class ReportMeta(
    val generatedAt: String,
    val from: String?,
    val to: String?,
    val days: Int,
    val nights: Int,
    val heartRateSamples: Int,
    val hrvWindows: Int,
    val activeDays: Int,
    val timeZone: String,
    val periodLabel: String,
    val profile: ReportProfile? = null,
)

@Serializable
data class ReportProfile(
    val heightCm: Double? = null,
    val weightKg: Double? = null,
)

/**
 * Une tuile de synthèse en tête de rapport.
 *
 * [status] pilote la couleur de la pastille. Les valeurs admises sont `good`, `warn`,
 * `serious`, `critical` et `neutral`. Le moteur de dessin ignore toute autre valeur et
 * retombe sur `neutral`.
 *
 * [value] et [sub] sont les **seuls** champs du modèle qui portent du texte déjà formaté.
 * Partout ailleurs le modèle garde ses nombres bruts et le moteur de dessin formate.
 * Les règles de typographie française s'appliquent ici, et elles s'appliquent de la même
 * façon dans les deux implémentations : virgule décimale, espace fine insaisissable pour
 * les milliers, heure d'horloge en `0h44`, durée en `5 h 48`, date en `23 juin 2025`.
 *
 * [value] ne répète jamais l'unité : elle va dans [unit]. `value = "54"` avec
 * `unit = "bpm"`, `value = "131/82"` avec `unit = "mmHg"`.
 *
 * [unit] ne vaut `null` que lorsque la valeur porte déjà sa notation en elle-même : une
 * durée `"5 h 48"`, une heure d'horloge `"0h44"`, un indice sans unité comme l'IMC.
 *
 * Sans donnée, [value] vaut `"—"` et [status] vaut `neutral`. Jamais un état inventé.
 */
@Serializable
data class ReportTile(
    val key: String,
    val label: String,
    val value: String,
    val unit: String? = null,
    val sub: String,
    val status: String,
)

/** Une valeur rattachée à un jour. */
@Serializable
data class DayValue(val date: String, val value: Double?)

/** Une valeur rattachée à un mois `AAAA-MM`. */
@Serializable
data class MonthValue(val month: String, val value: Double?)

/**
 * Une valeur rattachée à une étiquette libre : jour de semaine, tranche horaire, tranche de durée.
 *
 * Pour une série horaire, [label] est l'heure en clair, sans zéro devant et sans suffixe :
 * `"0"` à `"23"`. Le moteur de dessin la convertit en nombre pour placer le point ; un
 * `"00h"` donnerait `NaN` et effacerait la courbe. Le suffixe est ajouté à l'affichage.
 *
 * Pour un jour de semaine, [label] vaut `"lun."` à `"dim."`, dans cet ordre.
 *
 * [count] porte l'effectif quand il éclaire la valeur : le nombre de nuits d'une tranche
 * de durée, le nombre de jours mesurés derrière une moyenne.
 */
@Serializable
data class LabelValue(val label: String, val value: Double?, val count: Int? = null)

/**
 * Une corrélation de Pearson mesurée sur les jours appariés.
 *
 * [n] est le nombre de jours où les deux mesures existent. En dessous de
 * [CorrelationItem.MIN_PAIRS] jours, le coefficient n'est pas montré : il ne veut rien dire.
 */
@Serializable
data class CorrelationItem(
    val label: String,
    val r: Double,
    val n: Int,
) {
    companion object {
        const val MIN_PAIRS = 30
        const val STRONG = 0.30
    }
}

/**
 * Le texte du rapport, écrit par le LLM à partir des seuls agrégats.
 *
 * Cette partie est nulle tant que l'utilisateur n'a pas lancé d'analyse. Le rapport se
 * montre alors sans prose : les graphiques et les chiffres suffisent.
 */
@Serializable
data class ReportNarrative(
    val headline: String,
    val verdict: String,
    val sections: Map<String, SectionNarrative> = emptyMap(),
    val plan: List<PlanStep> = emptyList(),
)

@Serializable
data class SectionNarrative(
    val verdict: String,
    val points: List<String> = emptyList(),
)

@Serializable
data class PlanStep(
    val title: String,
    val body: String,
)

File: web/lib/report-model.js
/**
 * Contrat de données du rapport de santé — version JavaScript.
 *
 * Ce fichier est le miroir exact de
 * `app/src/main/java/com/kmt/healthanalyzer/domain/report/ReportModel.kt` et
 * `ReportSections.kt`. Le moteur de graphiques (`web/report/`) lit ce modèle et rien
 * d'autre. Il est utilisé tel quel par la version web et par la WebView Android.
 *
 * Toute modification ici doit être reportée côté Kotlin, et l'inverse. Les deux
 * plateformes doivent produire le même JSON, sinon les deux versions divergent.
 *
 * Conventions :
 * - Une date est une chaîne `AAAA-MM-JJ`. Un mois est une chaîne `AAAA-MM`.
 * - Une heure relative est exprimée en heures décimales autour de minuit : `-1.5` vaut
 *   22h30, `2.25` vaut 2h15. Sans cette convention, l'axe se coupe à minuit.
 * - Une valeur absente vaut `null`, jamais zéro. Un zéro se dessine ; un trou ne se
 *   dessine pas.
 *
 * ATTENTION : ce modèle porte des mesures quotidiennes. Il ne doit jamais partir vers un
 * fournisseur de LLM. La seule porte de sortie reste `lib/prompt.js`, qui n'envoie que
 * des agrégats hebdomadaires.
 *
 * Conventions que les deux implémentations doivent partager, sous peine de montrer des
 * chiffres différents pour les mêmes données. Le test de parité les vérifie.
 *
 * - **Étiquette horaire** : l'heure en clair, sans zéro devant et sans suffixe, `"0"` à
 *   `"23"`. Le moteur la convertit en nombre ; un `"00h"` donnerait `NaN`.
 * - **Jour de semaine** : `"lun."` à `"dim."`, dans cet ordre.
 * - **Stades de sommeil** : le dénominateur est la somme des quatre stades du mois, pas
 *   la durée de sommeil. Les quatre parts totalisent donc exactement 100.
 * - **Pas et étages par mois** : des moyennes quotidiennes, pas des totaux.
 * - **Exercice par mois** : des totaux du mois, pas des moyennes par séance.
 * - **Écart-type** : divisé par `n`, jamais par `n - 1`. Le rapport décrit un ensemble
 *   observé, il n'estime pas une population.
 * - **Médiane** : sur un effectif pair, la moyenne des deux valeurs centrales.
 * - **Heures relatives** : calculées à partir de la seconde du jour. Les millisecondes
 *   sont écartées : elles n'ont aucun sens sur une heure de coucher et ne servent qu'à
 *   faire diverger deux calculs. Ne pas arrondir à la minute non plus.
 * - **Codes Samsung non documentés** (ECG, apnée) : jamais traduits en libellé clinique.
 *   La forme est `"Résultat <code> — à lire dans Samsung Health Monitor"`.
 * - **Tuiles** : `value` et `sub` sont les seuls champs déjà formatés. Typographie
 *   française — virgule décimale, espace fine insaisissable pour les milliers, `0h44`
 *   pour une heure, `5 h 48` pour une durée, `23 juin 2025` pour une date. `value` ne
 *   répète jamais l'unité : elle va dans `unit` — `"54"` avec `"bpm"`, `"131/82"` avec
 *   `"mmHg"`. `unit` ne vaut `null` que si la valeur porte déjà sa notation : une durée
 *   `5 h 48`, une heure `0h44`, un indice sans unité comme l'IMC.
 *
 * @typedef {{date: string, value: number|null}} DayValue
 * @typedef {{month: string, value: number|null}} MonthValue
 * @typedef {{label: string, value: number|null, count?: number|null}} LabelValue
 * @typedef {{label: string, r: number, n: number}} CorrelationItem
 *
 * @typedef {{key: string, label: string, value: string, unit?: string|null,
 *            sub: string, status: 'good'|'warn'|'serious'|'critical'|'neutral'}} ReportTile
 *
 * @typedef {{generatedAt: string, from: string|null, to: string|null, days: number,
 *            nights: number, heartRateSamples: number, hrvWindows: number,
 *            activeDays: number, timeZone: string, periodLabel: string,
 *            profile: {heightCm: number|null, weightKg: number|null}|null}} ReportMeta
 *
 * @typedef {{date: string, hours: number, score: number|null, bedRel: number|null,
 *            wakeRel: number|null, efficiencyPercent: number|null, sessions: number}} SleepNightPoint
 * @typedef {{month: string, meanHours: number, medianHours: number,
 *            meanScore: number|null, nights: number}} SleepMonthStat
 * @typedef {{month: string, deep: number, light: number, rem: number, awake: number}} SleepStageMonth
 * @typedef {{nightly: SleepNightPoint[], monthly: SleepMonthStat[], dayOfWeek: LabelValue[],
 *            distribution: LabelValue[], stagesMonthly: SleepStageMonth[], kpi: object}} SleepSection
 *
 * @typedef {{month: string, resting: number|null, average: number|null}} HeartMonth
 * @typedef {{date: string, systolic: number, diastolic: number, pulse: number|null}} BloodPressurePoint
 * @typedef {{date: string, meanHeartRate: number|null, classification: number|null,
 *            classificationLabel: string}} EcgPoint
 *
 * @typedef {{month: string, sessions: number, minutes: number, calories: number|null}} ExerciseMonth
 * @typedef {{date: string, weightKg: number, bodyFatPercent: number|null,
 *            skeletalMuscleKg: number|null, bodyMassIndex: number|null,
 *            basalMetabolicRate: number|null}} BodyPoint
 * @typedef {{month: string, mean: number, percentAbove60: number}} StressMonth
 * @typedef {{month: string, mean: number, min: number}} Spo2Month
 *
 * @typedef {{headline: string, verdict: string,
 *            sections: Object<string, {verdict: string, points: string[]}>,
 *            plan: {title: string, body: string}[]}} ReportNarrative
 */

(function (global) {
  'use strict';

  /** Nombre minimal de jours appariés en dessous duquel une corrélation ne veut rien dire. */
  const MIN_CORRELATION_PAIRS = 30;

  /** Au-delà de ce coefficient, l'association mérite d'être signalée. */
  const STRONG_CORRELATION = 0.3;

  /** Les six sections du rapport, dans l'ordre où elles sont montrées. */
  const SECTION_KEYS = ['sleep', 'heart', 'activity', 'body', 'stress', 'breathing'];

  /** Les états admis pour la pastille d'une tuile. Toute autre valeur retombe sur `neutral`. */
  const TILE_STATUS = ['good', 'warn', 'serious', 'critical', 'neutral'];

  /**
   * Crée un modèle vide et valide.
   *
   * Le moteur de dessin sait montrer un rapport vide : il enlève les sections sans
   * données au lieu de dessiner des axes sans courbe.
   *
   * @returns {object} un modèle de rapport sans aucune mesure
   */
  function emptyReportModel() {
    return {
      meta: {
        generatedAt: new Date().toISOString(),
        from: null,
        to: null,
        days: 0,
        nights: 0,
        heartRateSamples: 0,
        hrvWindows: 0,
        activeDays: 0,
        timeZone: 'UTC',
        periodLabel: '',
        profile: null,
      },
      tiles: [],
      sleep: {
        nightly: [], monthly: [], dayOfWeek: [], distribution: [], stagesMonthly: [],
        kpi: { nights: 0, targetHours: 7.5, snoringNights: 0, snoringMeasuredNights: 0 },
      },
      heart: {
        monthly: [], restingDaily: [], hourly: [], hrvMonthly: [], hrvDaily: [],
        bloodPressure: [], ecg: [], kpi: { measuredDays: 0 },
      },
      activity: {
        stepsDaily: [], stepsRolling7: [], stepsMonthly: [], stepsDayOfWeek: [],
        exerciseMonthly: [], exerciseByKind: [], floorsMonthly: [],
        kpi: { exerciseSessions: 0, measuredDays: 0 },
      },
      body: { daily: [], kpi: { measures: 0 } },
      stress: {
        monthly: [], hourly: [], dayOfWeek: [], daily: [], vitalityDaily: [],
        kpi: { alerts: 0, measuredDays: 0 },
      },
      breathing: {
        spo2Monthly: [], spo2Daily: [], respiratoryDaily: [], skinTempDaily: [],
        kpi: { spo2Measures: 0, spo2Under90: 0 },
      },
      correlations: [],
      narrative: null,
    };
  }

  /**
   * Vérifie qu'un objet respecte le contrat, et renvoie la liste des manques.
   *
   * Cette fonction sert aux tests et au chargement d'un rapport exporté. Elle ne corrige
   * rien : elle nomme ce qui manque pour que l'erreur soit lisible.
   *
   * @param {object} model le modèle à vérifier
   * @returns {string[]} les problèmes trouvés, vide si le modèle est conforme
   */
  function validateReportModel(model) {
    const problems = [];
    if (!model || typeof model !== 'object') {
      return ['le modèle est absent ou n\'est pas un objet'];
    }
    if (!model.meta || typeof model.meta.generatedAt !== 'string') {
      problems.push('meta.generatedAt manque');
    }
    if (!Array.isArray(model.tiles)) {
      problems.push('tiles n\'est pas un tableau');
    }
    for (const key of SECTION_KEYS) {
      if (!model[key] || typeof model[key] !== 'object') {
        problems.push(`la section ${key} manque`);
      }
    }
    if (!Array.isArray(model.correlations)) {
      problems.push('correlations n\'est pas un tableau');
    }
    for (const tile of model.tiles || []) {
      if (!TILE_STATUS.includes(tile.status)) {
        problems.push(`la tuile ${tile.key} porte un état inconnu : ${tile.status}`);
      }
    }
    return problems;
  }

  global.HA = global.HA || {};
  global.HA.reportModel = {
    MIN_CORRELATION_PAIRS,
    STRONG_CORRELATION,
    SECTION_KEYS,
    TILE_STATUS,
    emptyReportModel,
    validateReportModel,
  };
}(typeof self !== 'undefined' ? self : this));
